dbt/include/snowflake/macros/adapters.sql

```diff
@@ -47,24 +47,90 @@
   {% do return(columns) %}
 {% endmacro %}
 
-{% macro snowflake__list_schemas(database) -%}
-  {# 10k limit from here: https://docs.snowflake.net/manuals/sql-reference/sql/show-schemas.html#usage-notes #}
-  {% set maximum = 10000 %}
-  {% set sql -%}
-    show terse schemas in database {{ database }}
-    limit {{ maximum }}
-  {%- endset %}
-  {% set result = run_query(sql) %}
-  {% if (result | length) >= maximum %}
-    {% set msg %}
-      Too many schemas in database {{ database }}! dbt can only get
-      information about databases with fewer than {{ maximum }} schemas.
-    {% endset %}
-    {% do exceptions.raise_compiler_error(msg) %}
-  {% endif %}
-  {{ return(result) }}
+{% macro snowflake__get_paginated_schemas_array(max_iter, max_results_per_iter, max_total_results, database, watermark) %}
```

Comment:
This code looks familiar, did we use pagination for a different query recently? If so, is there any way to reuse that, or to augment that so that both use cases can use it? The only piece that appears to be use case specific is `paginated_sql` and the error message. The former could be an argument to this macro and the latter could probably be made more generic (e.g. swap `schemas` for `objects`).

Reply:
Yep, you are remembering correctly @mikealfare !

> The solution is similar to that within https://github.com/dbt-labs/dbt-snowflake/pull/572, but for paginating the listing of **schemas** instead of paginating the listing of **relations**.